### PR TITLE
[Concurrency] Defensive copying of task locals in task groups, rather than crashing

### PR DESCRIPTION
**Pull request by ktoso**
This is a more advanced version of #73881 

This addresses a memory safety issue that task groups and task locals have faced since their introduction and has been previously guarded against by crashing rather than allowing the unsafety. The very specific situation this is about is the following:

```swift
      await withTaskGroup(of: Void.self) { group in
        TL.$two.withValue(2222) { // PREVIOUSLY: crash and warn against this usage pattern
          group.addTask {
            print("Survived, two: \(TL.two)")
          }
        }
```

The problem is that the withValue is effectively a push and a pop of the task local binding. The and **child tasks** refer directly to their parent's task local storage in order to minimize copying. This is fantastic because operations such as this:

```swift
TL.$a.withValue(...) {
TL.$b.withValue(...) {
TL.$c.withValue(...) {
TL.$d.withValue(...) { 
await withTaskGroup(of: Void.self) { group in
  group.addTask {}  // no copying, direct to parent reference (safe and good)
}
```

will cause zero copying. (good and safe)

However, the case of DIRECTLY wrapping an addTask with a task local binding like this: `TL.$two.withValue(2222) {  group.addTask {} }` is unsafe because the "pop" of the $two binding would happen BEFORE the child task gets to run to completion -- would would allow it to refer to released memory.

**Previous solution:**
We previously would crash whenever `withValue` was used inside a task group *at all*. Which led to problematic and unexpected crashes when more complex libraries which wrap code in task groups would call to user code or set task locals, causing people to work around issues: https://github.com/apple/swift-testing/pull/339 and a few more in server libraries.

Sometimes they may not be able to work around these incompatibilities though, leading to a not well composing structured task hierarchy (bad!)

**This solution:**
We instead now detect task locals pushed during the body of a task group. Then, as we `addTask` we defensively COPY them if directly in the parent task. This way the following situation is safe:

```swift
  await TL.$one.withValue(11) {
    await TL.$one.withValue(1111) {
      await withTaskGroup(of: Void.self) { group in

        TL.$two.withValue(2222) {
          group.addTask { 
          // task locals:
          // - copy 2222 item
          // - next reference to 1111
          // [skip] 11 no need to refer to it, since one already bound   
          }
        }
```

etc.

**TL;DR:**

- Instead of crashing in order to prevent bad memory access, we not handle it by copying.
- The situation where we copy is very rare and does not impact most of users: because the usual way is to bind around the task group
- It is simple to avoid the copying: move from `withValue { addTask {} }` to `addTask { withValue {} }` which causes no copying.

**Result:** Instead of crashing, we now can support this pattern - which sometimes comes up unbeknownst to actual developers, because they may be using libraries which wrap them in task groups and whatnot.

resolves rdar://127156303

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please test

File: stdlib/public/Concurrency/Executor.swift [hunk 1]
```diff
@@ -348,6 +348,17 @@ extension UnownedTaskExecutor: Equatable {
   }
 }
 
+@_unavailableInEmbedded
+@available(SwiftStdlib 6.0, *)
+extension UnownedTaskExecutor: Hashable {
```
Comment (FranzBusch) [inline stdlib/public/Concurrency/Executor.swift:353]: Should this be in here?

File: include/swift/ABI/TaskLocal.h [hunk 2]
```diff
@@ -101,20 +105,59 @@ class TaskLocal {
...
 
     Item *getNext() {
       return reinterpret_cast<Item *>(next & ~statusMask);
     }
 
+    void relinkNext(Item* nextOverride) {
+      fprintf(stderr, "[%s:%d](%s) this item = %p\n", __FILE_NAME__, __LINE__, __FUNCTION__, this);
+      fprintf(stderr, "[%s:%d](%s) NEXT = %p\n", __FILE_NAME__, __LINE__, __FUNCTION__, getNext());
+      fprintf(stderr, "[%s:%d](%s) storage = %p\n", __FILE_NAME__, __LINE__, __FUNCTION__, getStoragePtr());
+      fprintf(stderr, "[%s:%d](%s) next kind = %d\n", __FILE_NAME__, __LINE__, __FUNCTION__, getNextLinkType());
```
Comment (ktoso) [inline include/swift/ABI/TaskLocal.h:132]: TODO: cleanup all those prints

File: stdlib/public/Concurrency/Executor.swift (hunk 1 above)
Comment (ktoso) [reply stdlib/public/Concurrency/Executor.swift:353]: Thanks, nope :)

**Comment by ktoso** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
@swift-ci please test

File: stdlib/public/Concurrency/Actor.cpp [hunk 3]
```diff
@@ -378,7 +376,6 @@ bool swift_bincompat_useLegacyNonCrashingExecutorChecks() {
 
   return legacyMode;
 }
-#pragma clang diagnostic pop
```
Comment (ktoso) [inline stdlib/public/Concurrency/Actor.cpp:381]: was producing warnings, i had introduced this recently -- sorry for noise

**Comment by ktoso** [comment]
@swift-ci please test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please test

File: test/Concurrency/Runtime/async_task_locals_in_task_group_may_need_to_copy.swift [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  static var one: Int = 1
+  @TaskLocal
+  static var two: Int = 2
+  @TaskLocal
+  static var three: Int = 3
+}
+
+// ==== ------------------------------------------------------------------------
+
+func test() async {
```
Comment (FranzBusch) [inline test/Concurrency/Runtime/async_task_locals_in_task_group_may_need_to_copy.swift:20]: Can we add a test with multiple nested task groups inside child tasks? Just to make sure we don't copy above a `withTaskGroup` scope

Comment (ktoso) [reply test/Concurrency/Runtime/async_task_locals_in_task_group_may_need_to_copy.swift:20]: That’sa good test to add thx. We will copy those too, too defensively tbh. Preventing that is rough as we run out of space for some flags detecting this… maybe optimizing that copy away is a separate future follow uo

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
resolves https://github.com/apple/swift/issues/73217

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS

File: include/swift/ABI/TaskLocal.h [hunk 5]
```diff
@@ -44,6 +44,14 @@ class TaskLocal {
...
     IsParent = 0b01,
+    /// The task local binding was created inside the body of a `withTaskGroup`,
+    /// and therefore must either copy it, or crash when a child task is created
+    /// using 'group.addTask' and it would refer to this task local.
+    ///
+    /// Items of this kind must be copied by a group child task for access
+    /// safety reasons, as otherwise the pop would happen before the child task
+    /// has completed.
+    IsNextCreatedInTaskGroupBody = 0b10,
   };
```
Comment (rjmccall) [inline include/swift/ABI/TaskLocal.h:55]: You don't need to do this now, but it kind of sounds like these are flags instead of a single "kind".

File: include/swift/ABI/TaskLocal.h [hunk 6]
```diff
@@ -101,20 +109,55 @@ class TaskLocal {
...
     }
 
+    void relinkNext(Item* nextOverride) {
+      assert(!getNext() &&
+               "Can only relink task local item that was not pointing at anything yet");
+      assert(nextOverride->isNextLinkPointer() ||
+             nextOverride->isParentPointer() &&
+                 "Currently relinking is only done within a task group to "
+                 "avoid within-taskgroup next pointers; attempted to point at "
+                 "task local declared within task group body though!");
```
Comment (rjmccall) [inline include/swift/ABI/TaskLocal.h:139]: `A || B && C` is parsed as `A || (B && C)`.  That turns out to be logically equivalent in this case, but still, it's better to get into the habit of parenthesizing assertion conditions.

The assertion and behavior here is strange — the method name suggests that it's just changing the `next` pointer, but you're really assuming a stronger condition.  It feels like you can't really assert that condition properly, which to me is usually a sign that the operation isn't as general as I'm thinking it is.  That's okay; operations don't always generalize.  It just suggests that you ought to name this method something that indicates its special behavior, and maybe feel less bad about writing something that doesn't feel like a generic operation.

I assume what we're doing here is cloning the entries and then calling `relinkNext` on them.

File: stdlib/public/Concurrency/Task.cpp [hunk 7]
```diff
@@ -729,17 +729,16 @@ swift_task_create_commonImpl(size_t rawTaskCreateFlags,
...
-    assert(parent != nullptr && "creating a child task with no active task");
+  AsyncTask *parent = jobFlags.task_isChildTask() ? currentTask : nullptr;
+
+  if (group) {
+    assert(parent && "a task created in a group must be a child task");
+    // Add to the task group, if requested.
+    if (taskCreateFlags.addPendingGroupTaskUnconditionally()) {
+      assert(group && "Missing group");
+      swift_taskGroup_addPending(group, /*unconditionally=*/true);
+    }
```
Comment (rjmccall) [inline stdlib/public/Concurrency/Task.cpp:741]: This is a logically-unrelated cleanup?  Seems reasonable, just trying to find the connection.

File: include/swift/Runtime/Concurrency.h [hunk 8]
```diff
@@ -663,11 +663,23 @@ void swift_task_localValuePop();
...
+/// created in a task group. See task creation (swift_task_create_common) for
+/// a detailed discussion how and when this is used.
+///
+/// Its Swift signature is
+///
+/// \code
+/// func swift_task_localsCopyToTaskGroupChildTaskDefensively<Key>(AsyncTask* task)
+/// \endcode
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+void swift_task_localsCopyToTaskGroupChildTaskDefensively(AsyncTask* target);
```
Comment (rjmccall) [inline include/swift/Runtime/Concurrency.h:681]: I don't think this needs to be an exported runtime function.

**Review by rjmccall — COMMENTED** [review]
So, this is eagerly copying everything that was pushed inside of any task group, right?  That's actually subtly less than optimal in the case where there are multiple task groups active at once — we only really need to copy *up to the task group that the task is being created within*.  It's fine to do that for now, but let's not do anything that commits us to that in the long term, like making anything about this public ABI.  (This doesn't need to be public anyway — it's all internal to the concurrency library.)

File: include/swift/Runtime/Concurrency.h (hunk 8 above)
Comment (ktoso) [reply include/swift/Runtime/Concurrency.h:681]: I think we can managed without exposing it -- will remove (and remove _before_ getting this into 6.0)

**Comment by ktoso** [comment]
Thanks for the review John!

> So, this is eagerly copying everything that was pushed inside of any task group, right? That's actually subtly less than optimal in the case where there are multiple task groups active at once
Right, we can optimize this a bit more eventually, I didn't find a good way to do this yet, we'd have to store "which group it was created in" and I guess we'd probably either need to make task local 'item' larger or find a different trick, like another kind of record...

I'll make sure we don't lock ourselfs out of the optimization, following up now

File: include/swift/ABI/TaskLocal.h (hunk 5 above)
Comment (ktoso) [reply include/swift/ABI/TaskLocal.h:55]: Hm, lemme think if I can put these somewhere else 🤔

File: stdlib/public/Concurrency/Task.cpp (hunk 7 above)
Comment (ktoso) [reply stdlib/public/Concurrency/Task.cpp:741]: Yes, this bit is just cleanup -- same functionality.

File: include/swift/ABI/TaskLocal.h (hunk 6 above)
Comment (ktoso) [reply include/swift/ABI/TaskLocal.h:139]: I think it's a good point to rename this method, it is a very specific trick. When we detect the "next" task local item is "safe" as in "not the bad pattern where we have to copy" we immediately link to it rather than continuing our copying dance. I'll rename and see if I can clarify some more.

Fixing the assertion too, thank you for noticing that

File: include/swift/ABI/TaskLocal.h [hunk 9]
```diff
@@ -212,6 +258,8 @@ class TaskLocal {
     /// "pop" of the `B` value - it was spawned from a scope where only B was observable.
     void copyTo(AsyncTask *target);
 
+    void copyToOnlyOnlyFromCurrent(AsyncTask *target);
```
Comment (ktoso) [inline include/swift/ABI/TaskLocal.h:261]: renamed this in follow up https://github.com/apple/swift/pull/74089

File: include/swift/Runtime/Concurrency.h (hunk 8 above)
Comment (ktoso) [reply include/swift/Runtime/Concurrency.h:681]: removed in https://github.com/apple/swift/pull/74089

File: include/swift/ABI/TaskLocal.h (hunk 6 above)
Comment (ktoso) [reply include/swift/ABI/TaskLocal.h:139]: cleaned up in https://github.com/apple/swift/pull/74089
